src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -257,97 +344,108 @@ private CSharpSyntaxNode ParseExpressionOrPattern(bool forCase)
+                {
+                    designation = ParseSimpleDesignation();
+                    return true;
+                }
 
-                    if (type == null &&
-                        propertySubpattern == null &&
-                        designation == null &&
-                        subPatterns.Count == 1 &&
-                        subPatterns[0].NameColon == null &&
-                        subPatterns[0].Pattern is ConstantPatternSyntax cp)
-                    {
-                        // There is an ambiguity between a deconstruction pattern `(` pattern `)`
-                        // and a constant expression pattern than happens to be parenthesized.
-                        // We normalize to the parenthesized expression, and semantic analysis
-                        // will notice the parens and treat it whichever way is semantically sensible,
-                        // giving priority to its treatment as a constant expression.
-                        return _syntaxFactory.ParenthesizedExpression(openParenToken, cp.Expression, closeParenToken);
-                    }
+                return false;
+            }
 
-                    var node = _syntaxFactory.DeconstructionPattern(type, openParenToken, subPatterns, closeParenToken, propertySubpattern, designation);
-                    return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
+            bool looksLikeCast()
+            {
+                var resetPoint2 = this.GetResetPoint();
+                try
+                {
+                    return this.ScanCast(forPattern: true);
                 }
-
-                if (parsePropertySubpattern())
+                finally
                 {
-                    parseDesignation();
-                    var node = _syntaxFactory.PropertyPattern(type, propertySubpattern, designation);
-                    return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
+                    this.Reset(ref resetPoint2);
+                    this.Release(ref resetPoint2);
                 }
+            }
 
-                if (type != null && parseDesignation())
+            if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken && (type != null || !looksLikeCast()))
+            {
+                // It is possible this is a parenthesized (constant) expression.
+                // We normalize later.
+                ParseSubpatternList(out var openParenToken, out var subPatterns, out var closeParenToken, SyntaxKind.OpenParenToken, SyntaxKind.CloseParenToken);
+                //if (this.CurrentToken.Kind == SyntaxKind.EqualsGreaterThanToken)
+                //{
+                //    // Testers do the darndest things, in this case putting a lambda expression where a pattern is expected.
+                //    return null;
+                //}
```

Comment:
Please remove the commented out code, or add a PROTOTYPE comment with some details

Reply:
Recorded in [https://github.com/dotnet/roslyn/issues/23222](https://github.com/dotnet/roslyn/issues/23222).

---
In reply to: [156499914](https://github.com/dotnet/roslyn/pull/23209#discussion_r156499914) [](ancestors = 156499914)